Feature: Support for platform authentication on Windows.

TheKalpeshPawar opened the pull request:
**Ignore the first commit. The work has changed at the end.**
This commit, just like the last one, is bigger. 
This commit is bigger because I had to rewrite the sample app. Earlier, everything was placed inside a single composable function, `SampleAppNavigation`, so lifecycle management had become difficult after a certain point. 

The PlatformAuthenticationScreen (formerly DevieAuthScreen) and ChooseAuthOptionScreen are no longer part of the library, as they shouldn't be. The developer will have to provide their implementation for these screens according to their use case. Both screens are now directly present in the sample app as they should be.

The only UI components the library now holds are for passcode implementation, which is necessary as it is one of the features provided by the library. 

For review, I will just mention the concerned files and where the reviewer should look.

<h2>Main files from the library</h2>

- `WindowsHelloAuthenticator`: 
 This file contains logic for providing authentication functions for Windows, which in turn call the functions from the native C code.

- `PlatformAuthenticator` (for each platform):
This file contains code for ultimately calling the platform authenticator for registering and authenticating users.

- `PlatformAuthenticatorProvider`:
In analogy, this file is like a repository used for communicating with the DB provider. 
This file contains a class which is intended to be used by the developer will use in turn for use the platform authenticator.

- `PlatformAuthenticatorStatus`:
This file contains an enum class which gives different statuses of the platform authenticator, like BIOMETRIC_NOT_AVAILABLE, BIOMETRIC_NOT_SET, DEVICE_CREDENTIAL_SET, etc. 
These status values are returned in a Set from the *DeviceAuthenticatorStatus* function from the *PlatformAuthenticator* class

---
<h2> Main files from the Sample App </h2>

- `ChooseAuthOptionsScreen`  and  `ChooseAuthOptionScreenViewModel` :
 This screen shows the demo for selecting the app lock in case the developer chooses to provide both options, Passcode and PlatformAuthentication, separately for authentication. But there is no restriction on using them together. This screen calls the registerUser function via its view model.

- `PlatformAuthenticationScreen`  and  `PlatformAuthenticationScreenViewModel` : 
 This screen is what the user will see each time they open the app. Here they will verify themself. This screen calls the *authenticate* function via its view model.

- `SampleAppNavigation`: This file handles all the navigation of the app and also holds the implementation of the sample login and home screen.

- `App` :
I am not using DI in the sample app. So instances of viewmodel and datastore and created in this file and passed to the SampleAppNavigation composable.

---
<h2>Demo Videos</h2>

<h4>Windows Hello with Fingerprint:</h4>
This laptop doesn't have an IR camera, so it doesn't support Face ID, but if some computer does, then it will be supported.

https://github.com/user-attachments/assets/a946da2f-8365-43ad-9bb6-d8f58d0d117b


---
<h4>Windows Hello with only PIN support:</h4>
Most Windows computers may not have a fingerprint sensor, so this shows the PIN demo.


https://github.com/user-attachments/assets/cd210276-d78d-46fc-8fe1-dce62cc517e4

---
<h4>Unsupported platforms (Linux and Web, for example):</h4>
Linux, Web (JS and WASMJS), Mac and iOS are not supported yet by the library.

https://github.com/user-attachments/assets/e021fc18-5e74-4265-848b-ef408c55e158

---
<h4>Android:</h4>
Also shows what happens in case if the user has not set up a PIN or biometrics.
On Windows, the app won't redirect for setting up the authenticator. We can only show a prompt or a dialogue box.
The user will have to set up Windows Hello in Windows.


https://github.com/user-attachments/assets/e76fac62-6e68-4471-bd1e-a0e1e992a5a5

Commit:
Created user Registration.kt and Authentication.kt files in dummy server. These two file contain code for generating registration data and authentication data to send to authenticator and when authenticator send data back they parse it and verify it too.

Commit:
Completed Registration and Authentication functions. Now I will work on the library.

Commit:
Completed Registration and Authentication functions. Now I will work on the library.

Commit:
Registration and data saving is working fine now.

Commit:
Verification and registration calls both are working.

Commit:
Verification and registration calls both are working and saving data.
I am not double verifying the data. Because I ger errors while verifying them.

Commit:
Integrating WindowsHelloAuthenticator into library. Modifying library structure.

Commit:
Integrated windows hello logic into library. Sample is building and running properly for Android.

Commit:
Integrated windows hello logic into library. Sample is building and running properly for Android.

Commit:
Desktop app also works properly now. Fixed lifecycle issues with calling registration and verification function.

Commit:
Placed native code files with latest changes.

Commit:
After this commit, I am going to rewrite the library.

Commit:
Changes the working of PlatformAuthenticationProvider.kt. This commits has errors.

Commit:
Updates data types in PlatformAuthenticator in all source sets and ths changes logic a little accordingly.

Commit:
Moved DeviceAuthScreen.kt and ChooseAuthOptionScreen into sample module as they are not supposed to the part of the library.

Commit:
Sample app works now.

Commit:
Some required changes in coroutine context which now prevents sample from crashing in authentication flow.

TheKalpeshPawar committed:
Fixed error message for unsupported platforms.

TheKalpeshPawar committed:
Android Biometric needs to run on Main thread, so now it will.

Commit:
Downgraded JVM version for android and desktop to 17 from 21.

Commit:
Updated version for some dependencies and removed 3 unused unnecessary dependencies

niyajali commented on mifos-passcode-cmp/build.gradle.kts:
```diff
@@ -82,6 +80,15 @@ kotlin {
+            implementation(libs.multiplatform.settings.serialization)
+            implementation(libs.multiplatform.settings.coroutines)
+
+            implementation("net.java.dev.jna:platform:3.5.2")
```
Move all of this library to version catalog

niyajali commented on mifos-passcode-cmp/build.gradle.kts:
```diff
@@ -104,6 +111,18 @@ kotlin {
                 implementation(compose.foundation)
                 implementation(libs.kotlinx.coroutines.swing)
+
+                implementation("com.webauthn4j:webauthn4j-core:0.29.2.RELEASE")
```
These too

TheKalpeshPawar replied:
Okay sir. Sir, I have one more commit that provides documentation for some functions and classes. Should I push in into this pr or create a seperate PR after this ?

niyajali commented on gradle/libs.versions.toml:
```diff
@@ -1,7 +1,7 @@
+agp = "8.9.3"
 koinAnnotations = "2.0.0"
-kotlin = "2.1.20"
+kotlin = "2.1.21"
```
Update the `devToolsKsp` version according to kotlin

niyajali commented on mifos-passcode-cmp/build.gradle.kts:
```diff
@@ -33,9 +32,8 @@ kotlin {
-            kotlinOptions.jvmTarget = "1.8"
-        }
+        compilerOptions{
+            jvmTarget.set(JvmTarget.JVM_21)        }
```
Use a lesser JVM version if possible to support this library in a wide range like 17, 11 etc

niyajali commented on mifos-passcode-cmp/build.gradle.kts:
```diff
@@ -145,6 +167,20 @@ kotlin {
 
+compose.desktop{
+    application {
+        mainClass = "com.mifos.passcode.auth.deviceAuth.PlatformAuthenticatorKt" // Or your actual mainKt class
```
remove all of this block comment(`// Or your actual mainKt class`) from all file instead put into top of the property or methods like this

```
 // Or your actual mainKt class
mainClass = "com.mifos.passcode.auth.deviceAuth.PlatformAuthenticatorKt"
```

niyajali commented on mifos-passcode-cmp/src/androidMain/kotlin/auth/deviceAuth/PlatformAuthenticator.kt:
```diff
@@ -7,12 +7,11 @@ import android.os.Build
-import androidx.biometric.BiometricManager.Authenticators.BIOMETRIC_STRONG
-import androidx.biometric.BiometricManager.Authenticators.BIOMETRIC_WEAK
-import androidx.biometric.BiometricManager.Authenticators.DEVICE_CREDENTIAL
+import androidx.biometric.BiometricManager.Authenticators.*
```
Remove all wildcard(*) imports, also disable in your IDE, and enable auto import

niyajali commented on mifos-passcode-cmp/src/androidMain/kotlin/auth/deviceAuth/PlatformAuthenticator.kt:
```diff
@@ -126,21 +99,24 @@ actual class PlatformAuthenticator private actual constructor() {
             .setSubtitle("Unlock using your PIN, Password, Pattern, Face or Fingerprint")
             .setAllowedAuthenticators(
-                if(apiLevel>29) BIOMETRIC_STRONG or DEVICE_CREDENTIAL
+                if (apiLevel > 29) BIOMETRIC_STRONG or DEVICE_CREDENTIAL
```
where does this  `apiLevel ` varriable defined

TheKalpeshPawar replied:
Hey @niyajali I need help. I made a mistake while pushing the commit for version catalogue changes.
I rebased while pushing, and now the commits look messy. I had some commits which I wasn't going to push right now.

TheKalpeshPawar replied:
apiLevel variable is in the same file, sir.

TheKalpeshPawar committed:
Moved some libraries  directly used in build.gradle.kts into version catalog.

TheKalpeshPawar committed:
Moved some libraries  directly used in build.gradle.kts into version catalog. Optimized imports.

Commit:
Authentication cancelled message was not showing on android source set. Now it will.

Commit:
Resolved potential memory management issues. Also resolved some issues in sample app which which be present due to viewmodel being initialized directly and thus causing crashes.

Commit:
Renaming 'mock_server' directory to 'mockServer'

Commit:
Provided a new way to inject Activity and Context.

Commit:
Deleting unnecessary files.

Commit:
Removed all println statements used for debugging purpose.

TheKalpeshPawar committed:
PlatformAuthenticationProvider and PlatformAvailableAuthenticationOption both can be now provided using CompositionLocal.

niyajali commented:
@biplab1 @HekmatullahAmin Can you guys suggest some changes on this PR

niyajali commented:
@HekmatullahAmin help him to set up Spotless & Detekt on this project

niyajali commented:
@biplab1 @HekmatullahAmin Also, can you guys suggest how we can modularize this library, check out other popular **kotlinx** or **square** libraries and come up with a solution for the same

TheKalpeshPawar committed:
PlatformAuthenticationProvider.kt was now automatically updates _currentAuthOptions status when initialized. The function for manually updating is value is not needed now.
I have also applied changes suggested by detekt plugin on files I worked on today. For other files I will apply them later.

HekmatullahAmin commented:
Hi @TheKalpeshPawar ,

As per what Niyaj mentioned, we're planning to transition this project into a multi-module Kotlin Multiplatform architecture later on. To make our tooling setup scalable and reusable across modules, they created custom Gradle convention plugins for Detekt and Spotless, you follow the same pattern we use in other Mifos projects like Kotlin-Multiplatform-Project.

some things might need to be changed handle it accordingly

1. build-logic Setup
1.1 Create build-logic/settings.gradle.kts
```
dependencyResolutionManagement {
    repositories {
        google()
        mavenCentral()
        gradlePluginPortal()
    }
    versionCatalogs {
        create("libs") {
            from(files("../gradle/libs.versions.toml"))
        }
    }
}

rootProject.name = "build-logic"
include(":convention")
```
1.2 Create build-logic/convention/build.gradle.kts
```
plugins {
    `kotlin-dsl`
}

group = "org.convention.buildlogic"

java {
    sourceCompatibility = JavaVersion.VERSION_17
    targetCompatibility = JavaVersion.VERSION_17
}

dependencies {
    compileOnly(libs.detekt.gradlePlugin)
    compileOnly(libs.spotless.gradle)
}

gradlePlugin {
    plugins {
        register("detektConvention") {
            id = "org.convention.detekt.plugin"
            implementationClass = "DetektConventionPlugin"
        }
        register("spotlessConvention") {
            id = "org.convention.spotless.plugin"
            implementationClass = "SpotlessConventionPlugin"
        }
    }
}
```
1.3 Create DetektConventionPlugin.kt
```
package org.convention

import org.gradle.api.Plugin
import org.gradle.api.Project
import io.gitlab.arturbosch.detekt.extensions.DetektExtension
import io.gitlab.arturbosch.detekt.Detekt
import org.gradle.kotlin.dsl.configure
import org.gradle.kotlin.dsl.dependencies
import org.gradle.kotlin.dsl.named

class DetektConventionPlugin : Plugin<Project> {
    override fun apply(target: Project) = with(target) {
        pluginManager.apply("io.gitlab.arturbosch.detekt")

        extensions.configure<DetektExtension> {
            buildUponDefaultConfig = true
            config.setFrom(rootProject.file("config/detekt/detekt.yml"))
        }

        tasks.named<Detekt>("detekt") {
            jvmTarget = "17"
            reports {
                xml.required.set(true)
                html.required.set(true)
            }
        }

        dependencies {
            "detektPlugins"("io.gitlab.arturbosch.detekt:detekt-formatting:${libs.versions.detekt.get()}")
        }
    }
}
```
1.4 Create SpotlessConventionPlugin.kt
```
package org.convention

import com.diffplug.gradle.spotless.SpotlessExtension
import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.kotlin.dsl.configure

class SpotlessConventionPlugin : Plugin<Project> {
    override fun apply(target: Project) = with(target) {
        pluginManager.apply("com.diffplug.spotless")

        extensions.configure<SpotlessExtension> {
            kotlin {
                target("**/*.kt")
                targetExclude("**/build/**")
                ktlint("0.48.2").editorConfigOverride(mapOf("android" to "true"))
                licenseHeaderFile(rootProject.file("spotless/copyright.kt"))
            }

            format("kts") {
                target("**/*.kts")
                licenseHeaderFile(rootProject.file("spotless/copyright.kts"))
            }
        }
    }
}
```
📁 2. Detekt Configuration
Create config/detekt/detekt.yml
```
copy it from other projects the rules are defiined their
```
📁 3. Spotless Headers
Create: copy the text also from other projects

spotless/copyright.kt
```
/*
 * Copyright (c) 2025.
 * All rights reserved.
 */
 ```
spotless/copyright.kts
```
/*
 * Copyright (c) 2025.
 * All rights reserved.
 */
 ```
 
🔧 4. Root Project Setup
settings.gradle.kts
```
pluginManagement {
    includeBuild("build-logic")
    repositories {
        google()
        mavenCentral()
        gradlePluginPortal()
    }
}

rootProject.name = "your-project-name"
include(":your-main-module") // or :app etc.
```
build.gradle.kts
```
plugins {
    id("org.convention.detekt.plugin")
    id("org.convention.spotless.plugin")
}
```
🧩 5. libs.versions.toml (Add Versions)
In gradle/libs.versions.toml: copy it from other projects the version whatever it is
```
[versions]
detekt = "1.23.8"
spotlessVersion = "6.22.0"

[libraries]
detekt-gradlePlugin = { group = "io.gitlab.arturbosch.detekt", name = "detekt-gradle-plugin", version.ref = "detekt" }
detekt-formatting = { group = "io.gitlab.arturbosch.detekt", name = "detekt-formatting", version.ref = "detekt" }
spotless-gradle = { group = "com.diffplug.spotless", name = "spotless-plugin-gradle", version.ref = "spotlessVersion" }
```

✅ 6. Apply to Your Main KMP Module
In your module’s build.gradle.kts (e.g., :shared, :app, etc.):
```
plugins {
    kotlin("multiplatform")
    id("org.convention.detekt.plugin")
    id("org.convention.spotless.plugin")
}
```
🧪 Done! Now Try:
If any issues appear, fix them accordingly
```
./gradlew detekt
./gradlew spotlessCheck
./gradlew spotlessApply
```

niyajali commented:
@HekmatullahAmin First review this PR and make it mergeable and find some time to connect with him and setup multi-module architecture and then setup deket & spotless in the whole project

HekmatullahAmin commented on mifos-passcode-cmp/src/desktopMain/nativeCode/WindowsHelloAuthenticator.c:
```diff
@@ -0,0 +1,82 @@
+    printf(verificationData->challenge);printf("\n");
+    printf(verificationData->origin);printf("\n");
+    printf(verificationData->rpId);printf("\n");
+    printf("Receiver credential Id: \n");
```
I believe you are putting these `printf` lines for debugging however for production these need to be removed. Code you also put `TODO` so that these need to be delete sth like // TODO these are for debuggin and should be removed later

HekmatullahAmin replied:
@TheKalpeshPawar brother you request for review but you still didn't implemented what is asked.
And also yesterday I mentioned when someone request you a change and you have different idea that why it can't be done give them a respone and say why.

HekmatullahAmin commented on mifos-passcode-cmp/src/commonMain/kotlin/auth/deviceAuth/PlatformAuthenticationProvider.kt:
```diff
@@ -1,47 +1,90 @@
+    val authenticatorStatus = _authenticatorStatus.asStateFlow()
 
     init {
         _authenticatorStatus.value = deviceAuthenticatorStatus()
```
You already set the _authenticatorStatus in its declaration and again in init. One is enough

HekmatullahAmin commented on sample/composeApp/src/commonMain/kotlin/com/mifos/passcode/sample/chooseAuthOption/ChooseAuthOptionScreen.kt:
```diff
@@ -0,0 +1,286 @@
+                    icon = Icons.Default.People,
+                    onSelect = {
+                        optionSet[0] = false
+                        optionSet[1] = true
```
Improve State for Auth Option Selection
sth like:
```
enum class AuthOption { None, DeviceLock, MifosPasscode }

var selectedOption by rememberSaveable { mutableStateOf(AuthOption.None) }
```

```
selected = selectedOption == AuthOption.DeviceLock
onSelect = { selectedOption = AuthOption.DeviceLock }
```

HekmatullahAmin commented on sample/composeApp/src/commonMain/kotlin/com/mifos/passcode/sample/chooseAuthOption/ChooseAuthOptionScreenViewmodel.kt:
```diff
@@ -0,0 +1,54 @@
+
+    private val _registrationResult = Channel<RegistrationResult?>()
+    val registrationResult = _registrationResult.receiveAsFlow()
+
```
Is there any reason you use Channel in my point of view u can use `MutableStateFlow` and you can easilty assing it and collect it.

```
 private val _registrationResult = MutableStateFlow<RegistrationResult?>(null)
    val registrationResult: StateFlow<RegistrationResult?> = _registrationResult
```
```
_registrationResult.value = provider.registerUser(
                userId,
                userEmail,
                displayName
            )
```

HekmatullahAmin commented on sample/composeApp/src/commonMain/kotlin/com/mifos/passcode/sample/chooseAuthOption/ChooseAuthOptionScreenViewmodel.kt:
```diff
@@ -0,0 +1,54 @@
+        userEmail: String = "",
+        displayName: String = ""
+    ) {
+        viewModelScope.launch(Dispatchers.Main) {
```
when we use `viewmodelScope.launch` by default it runs in Main thread no need to reannounce it:

HekmatullahAmin replied:
you still have this @TheKalpeshPawar

HekmatullahAmin commented on sample/composeApp/src/commonMain/kotlin/com/mifos/passcode/sample/platformAuthentication/AuthenticationScreen.kt:
```diff
@@ -0,0 +1,189 @@
+
+    val authenticatorStatus by platformAuthenticationProvider.authenticatorStatus.collectAsState()
+
+    println("Authenticator status: $authenticatorStatus")
```
same here these `println` should be removed if you keep it for now for debuggin just also add a `TODO` on top of it in future to remove

HekmatullahAmin commented on sample/composeApp/src/commonMain/kotlin/com/mifos/passcode/sample/platformAuthentication/AuthenticationScreen.kt:
```diff
@@ -0,0 +1,189 @@
+    navController: NavController,
+) {
+    val verificationResult = authenticationScreenViewModel.authenticationResult.collectAsState()
+
```
in mobile the `collectAsStateWithLifecycle()` is preffered than `collectAsState` so use that

HekmatullahAmin commented on sample/composeApp/src/commonMain/kotlin/com/mifos/passcode/sample/platformAuthentication/AuthenticationScreenViewModel.kt:
```diff
@@ -0,0 +1,43 @@
+
+    fun authenticateUser(appName: String, platformAuthenticationProvider: PlatformAuthenticationProvider) {
+        _isLoading.value = true
+        viewModelScope.launch(Dispatchers.Main) {
```
Use viewModelScope.launch Without Explicit Dispatcher

HekmatullahAmin commented on sample/composeApp/src/commonMain/kotlin/com/mifos/passcode/sample/platformAuthentication/AuthenticationScreen.kt:
```diff
@@ -0,0 +1,189 @@
+    LaunchedEffect(
+        verificationResult.value,
+    ) {
+        this.launch {
```
`LaunchedEffect` already launches a `coroutine`, so calling `this.launch {}` inside it is redundant

biplab1 commented on mifos-passcode-cmp/src/commonMain/kotlin/auth/deviceAuth/PlatformAuthenticationProvider.kt:
```diff
@@ -1,47 +1,90 @@
+                    displayName
+                )
+            } catch (e: Exception) {
+                RegistrationResult.Error("Registration failed: ${e.message}")
```
Using `e.message` directly might return `null`, causing a null string. A safer alternative could be: 
`RegistrationResult.Error("Registration failed: ${e.message ?: "Unknown error"}")`

Also, please use Kermit logger for logging the exception which can help with debugging:
`Logger.e(e) { "Registration failed for user: $userName, email: $emailId" }`

Likewise, ensure proper exception handling in other classes or functions where exceptions may occur.

biplab1 commented on mifos-passcode-cmp/src/commonMain/kotlin/crypto/CipherUtil.kt:
```diff
@@ -1,5 +1,5 @@
-package com.mifos.passcode.crypto
-
+//package com.mifos.passcode.crypto
+//
```
Could you please tell the reason for commenting this block of code?

biplab1 commented on sample/composeApp/src/commonMain/kotlin/com/mifos/passcode/sample/chooseAuthOption/ChooseAuthOptionScreen.kt:
```diff
@@ -0,0 +1,286 @@
+                                platformAuthenticationProvider,
+                                "thekalpeshpawar",
+                                "pawarkalpesh@proton.me",
+                                "Kalpesh Pawar"
```
Consider avoiding hardcoded values for userName, emailId, and displayName in the registration call. It would be better to extract these into constants or derive them from UI/user input. This will improve maintainability and avoid accidental production use of test/demo credentials.

biplab1 commented on mifos-passcode-cmp/src/desktopMain/kotlin/mockServer/models/UserAuthentication.kt:
```diff
@@ -0,0 +1,177 @@
+////        }
+////
+////    }
+////
```
May I know why this large section of code is commented out?

biplab1 commented on mifos-passcode-cmp/src/desktopMain/kotlin/mockServer/models/UserRegistration.kt:
```diff
@@ -0,0 +1,182 @@
+//            return Pair(false, null)
+//        }
+//
+//        val actualCollectedClientDataBytes = getCollectdClientDataBytes(origin, type.value, challenge)
+//
+//        val mOrigin = Origin(origin)
+//        val mChallenge = DefaultChallenge(challenge)
+//
+//        val collectedClientData = CollectedClientData(
+//            type,
+//            mChallenge,
+//            mOrigin,
+//            false,
+//            null
+//        )
+//
```
Same here

biplab1 reviewed (commented):
I've noticed large blocks of commented-out code across multiple files in this PR. Could you please clarify their purpose? If they're no longer needed, it would be cleaner to remove them to improve readability and reduce clutter.

This comment applies to similar cases elsewhere in the code as well. Let me know if there's a plan to reintroduce or refactor this code later—adding a short TODO or context comment might help in that case.

TheKalpeshPawar committed:
Wrote documentation for necessary classes and functions.

TheKalpeshPawar committed:
Put temporary documentation on some classes and function that are unused for now. Also removed some files that are no longer used.

TheKalpeshPawar committed:
Dependency cleanup.

biplab1 commented on mifos-passcode-cmp/src/desktopMain/kotlin/mockServer/Requests.kt:
```diff
@@ -0,0 +1,238 @@
+    fun getAuthenticatorDataBytes(): ByteArray? {
+        if (authenticatorDataBytes == null || authenticatorDataLength <= 0) {
+            println("Null AuthenticatorDataBytes")
+            return null
```
Instead of returning a null, prefer returning an empty ByteArray for absent data, or use a sealed/result type for clearer error handling.

biplab1 commented on mifos-passcode-cmp/src/desktopMain/kotlin/mockServer/Requests.kt:
```diff
@@ -0,0 +1,238 @@
+            println("Null AuthenticatorDataBytes")
+            return null
+        }
+        return authenticatorDataBytes!!.getByteArray(0, authenticatorDataLength)
```
Double-bang (!!) usage is unnecessary after the null check, but it's still better to avoid if possible.

biplab1 commented on mifos-passcode-cmp/src/desktopMain/kotlin/mockServer/Requests.kt:
```diff
@@ -0,0 +1,238 @@
+        return authenticatorDataBytes!!.getByteArray(0, authenticatorDataLength)
+    }
+
+    fun getSignatureDataBytes(): ByteArray? {
+        if (signatureDataBytes == null || signatureDataBytesLength <= 0) {
+            println("Null signatureDataBytes")
+            return null
+        }
+        return signatureDataBytes!!.getByteArray(0, signatureDataBytesLength)
+    }
+
+    fun getUserHandleBytes(): ByteArray? {
+        if (userHandle == null || userHandleLength <= 0) {
+            println("Null user handle")
+            return null
+        }
+        return userHandle!!.getByteArray(0, userHandleLength)
+    }
```
Prefer returning an empty ByteArray instead of null for absent data, and avoid using !! even after a null check—consider using a sealed or result type for clearer and safer error handling.

biplab1 commented on mifos-passcode-cmp/src/desktopMain/kotlin/mockServer/Requests.kt:
```diff
@@ -0,0 +1,238 @@
+            ")"
+    }
+
+    fun getAttestationObjectBytes(): ByteArray? {
+        if (attestationObjectBytes == null || attestationObjectLength <= 0 || getAuthenticationResult() != WindowsAuthenticationResponse.SUCCESS) {
+            return null
+        }
+        val attstObj = attestationObjectBytes!!.getByteArray(0, attestationObjectLength)
+
+        return attstObj
+    }
+
+    fun getCredentialIDBytes(): ByteArray? {
+        if (credentialIdBytes == null || credentialIdLength <= 0) {
+            return null
+        }
+        return credentialIdBytes!!.getByteArray(0, credentialIdLength)
+    }
```
same here, please review.

TheKalpeshPawar committed:
Using sealed class to handle error instead of null values.

TheKalpeshPawar committed:
Update utils.c

TheKalpeshPawar committed:
Update PlatformAuthenticator.kt docs

TheKalpeshPawar committed:
Removed explicit mention on Main Dispatcher from AuthenticationScreenViewModel.kt and ChooseAuthOptionScreenViewmodel.kt.

TheKalpeshPawar committed:
Removed printf satements from the WindowsHelloAuthenticator.c file

TheKalpeshPawar committed:
Update registration.c

Remode printf statements from registration.c file

TheKalpeshPawar committed:
Update userVerification.c

Remoed printf statemnts.

TheKalpeshPawar committed:
Update utils.c

Remoed fprintf satements

TheKalpeshPawar committed:
Update mock_server.c

Remode fprintf and printf statements

biplab1 commented on gradlew:
This file is empty, can be removed

biplab1 commented on mifos-passcode-cmp/src/desktopMain/kotlin/mockServer/models/UserAuthentication.kt:
```diff
@@ -0,0 +1,204 @@
+// //            return Pair(true, windowsAuthenticatorDataToSave)
+// //        }catch (e: VerificationException){
+// //            e.printStackTrace()
+// //            println("Personalized message: ")
+// //            println(e.message)
+// //            println(e.cause)
+// //            println(e.localizedMessage)
+// //
+// //            return Pair(false, null)
+// //        }
+// //
+// //    }
+// //
+// //}
```
It looks like the comments have been applied twice in the file.

TheKalpeshPawar committed:
Update UserAuthentication.kt

Removed double comments for the same line.
